lib/service_base_url_test_kit/service_base_url_validate_group.rb

```diff
@@ -152,37 +291,45 @@ def skip_message
       )
 
       run do
-        bundle_response = if service_base_url_bundle.blank?
-                            load_tagged_requests('service_base_url_bundle')
-                            skip skip_message if requests.length != 1
-                            requests.first.response_body
-                          else
-                            service_base_url_bundle
-                          end
-        skip_if bundle_response.blank?, 'No Bundle response was provided'
+        bundle_resource = scratch[:bundle_resource]
 
-        assert_valid_json(bundle_response)
-        bundle_resource = FHIR.from_contents(bundle_response)
+        skip_if bundle_resource.blank?, 'No Bundle response was provided'
 
         skip_if bundle_resource.entry.empty?, 'The given Bundle does not contain any resources'
 
-        assert_valid_bundle_entries(bundle: bundle_resource,
-                                    resource_types: {
-                                      Endpoint: nil
-                                    })
-
-        endpoint_ids =
+        endpoint_resources =
           bundle_resource
             .entry
             .map(&:resource)
             .select { |resource| resource.resourceType == 'Endpoint' }
-            .map(&:id)
 
-        endpoint_ids.each do |endpoint_id|
+        endpoint_resources.each do |endpoint|
+          resource_is_valid?(resource: endpoint)
+
+          endpoint_id = endpoint.id
           endpoint_referenced_orgs = find_referenced_org(bundle_resource, endpoint_id)
-          assert !endpoint_referenced_orgs.empty?,
-                 "Endpoint with id: #{endpoint_id} does not have any associated Organizations in the Bundle."
+          next unless endpoint_referenced_orgs.empty?
+
+          add_message('error', %(
+            Endpoint with id: #{endpoint_id} does not have any associated Organizations in the Bundle.
+          ))
+        end
+
+        error_messages = messages.select { |msg| msg[:type] == 'error' }
+        non_error_messages = messages.reject { |msg| msg[:type] == 'error' }
+
+        @messages = []
+        @messages += error_messages.first(20) unless error_messages.empty?
+        @messages += non_error_messages.first(50) unless non_error_messages.empty?
```

Comment:
The validator adds the messages by sticking them in `@messages`. Those messages are then persisted by the test runner after the test finishes executing, so that part seems fine.

I don't know why we'd want to allow more info/warning messages than error messages, though, as error messages are more important. It's probably also a good idea to put the message saying that only the first xyz messages are being shown first rather than last.

I also think a better solution would be to check if the limit has been reached after each iteration, and then breaking once the limit has been reached. That would prevent performing useless validations and also ensure that all of the messages for each resource validated are kept.

Reply:
I was also thinking that would be a better solution, but the issue with that approach is if there are a ton of info/warning messages. The problem I realized with that approach is that if there are a ton of info/warning messages then you might hit the limit of messages and break before you validate a resource that would have had an error and you wouldn't see it, whereas with this approach if there is an error, it will be caught.